core/controllers/acl_decorators.py

```diff
@@ -2318,11 +2338,47 @@ def test_can_access(self, topic_name, **kwargs):
         Raises:
             PageNotFoundException: The given page cannot be found.
         """
-        topic_name = python_utils.url_unquote_plus(topic_name)
-        topic = topic_fetchers.get_topic_by_name(topic_name)
+        lowercase_abbrev_topic_name = (
+            python_utils.url_unquote_plus(abbrev_topic_name).replace('-', ' '))
+        topic = topic_fetchers.get_topic_by_abbreviated_name(
+            lowercase_abbrev_topic_name)
+        classroom_name = kwargs.pop('classroom_name', None)
 
         if topic is None:
-            raise self.PageNotFoundException
+            handle_redirection(
+                self, '/%s' % classroom_name,
+                self.GET_HANDLER_ERROR_RETURN_TYPE)
+            return
+        if classroom_name not in [None, constants.CLASSROOM_STATUS_STAGING]:
+            lowercase_classroom_name = (
+                python_utils.url_unquote_plus(
+                    classroom_name).replace('-', ' '))
+            topic_ids_from_classroom_pages = (
+                config_domain.TOPIC_IDS_FOR_CLASSROOM_PAGES.value)
+            topic_is_assigned_classroom = False
+            for classroom_dict in topic_ids_from_classroom_pages:
+                if topic.id in classroom_dict['topic_ids']:
+                    if (
+                            lowercase_classroom_name ==
+                            classroom_dict['name'].lower()):
+                        topic_is_assigned_classroom = True
+                        break
+                    else:
+                        handle_redirection(
```

Comment:
As a note -- these "error" redirections should not be permanent. Do not permanently redirect URLs to an error page. Only have permanent redirections from old URLs to their corresponding new URLs.

Reply:
If the URL field values passed are invalid, you redirect "upwards", right? That redirection should not be permanent. As an example, someone may accidentally enter a URL for a chapter that is yet to be written, so the URL may actually become valid someday. You don't want the URL to end up being invalid forever because of a 301 redirect.